File: rules/gitlab/java/xml/test-XsltTransform.java
// License: LGPL-3.0 License (c) find-sec-bugs
package xml;


import org.apache.commons.io.IOUtils;

import javax.xml.transform.Source;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerException;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.stream.StreamResult;
import javax.xml.transform.stream.StreamSource;
import java.io.FileInputStream;
import java.io.IOException;

public class XsltTransform {

    public static final String FOLDER = "/testcode/xsl/";

    public void xslt1SafeStaticResource() throws TransformerException {
        TransformerFactory factory = TransformerFactory.newInstance();
        Source xslt = new StreamSource(getClass().getResourceAsStream(FOLDER + "xsl_safe.xsl"));
        Transformer transformer = factory.newTransformer(xslt);

        Source text = new StreamSource(getClass().getResourceAsStream(FOLDER + "input.xml"));
        transformer.transform(text, new StreamResult(System.out));
    }

    public void xslt2UnsafeResource(String input) throws TransformerException {
        TransformerFactory factory = TransformerFactory.newInstance();
        Source xslt = new StreamSource(getClass().getResourceAsStream(input));
        Transformer transformer = factory.newTransformer(xslt);

        Source text = new StreamSource(getClass().getResourceAsStream(FOLDER + "input.xml"));
        transformer.transform(text, new StreamResult(System.out));
    }

    public void xslt3UnsafeResource(String input) throws TransformerException {
        TransformerFactory factory = TransformerFactory.newInstance();
        Source xslt = new StreamSource(getClass().getResourceAsStream(FOLDER + input));
        Transformer transformer = factory.newTransformer(xslt);

        Source text = new StreamSource(getClass().getResourceAsStream(FOLDER + "input.xml"));
        transformer.transform(text, new StreamResult(System.out));
    }

    public void xslt4UnsafeResource(String input) throws TransformerException {
        TransformerFactory factory = TransformerFactory.newInstance();
        FileInputStream in = null;
        try {
            in = new FileInputStream(FOLDER + input);
            Source xslt = new StreamSource(in);
            Transformer transformer = factory.newTransformer(xslt);

            Source text = new StreamSource(getClass().getResourceAsStream(FOLDER + "input.xml"));
            transformer.transform(text, new StreamResult(System.out));
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            IOUtils.closeQuietly(in);
        }
    }

    public void xslt5SafeResource() throws TransformerException {
        String input = "/safe.xsl";
        TransformerFactory factory = TransformerFactory.newInstance();
        FileInputStream in = null;
        try {
            in = new FileInputStream(FOLDER + input);
            Source xslt = new StreamSource(in);
            Transformer transformer = factory.newTransformer(xslt);

            Source text = new StreamSource(getClass().getResourceAsStream(FOLDER + "input.xml"));
            transformer.transform(text, new StreamResult(System.out));
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            IOUtils.closeQuietly(in);
        }
    }
}

File: rules/gitlab/scala/xml/test-XsltTransform.scala
// License: LGPL-3.0 License (c) find-sec-bugs
package xml

import org.apache.commons.io.IOUtils

import javax.xml.transform.{Source, Transformer, TransformerException, TransformerFactory, stream}
import javax.xml.transform.stream.StreamResult
import javax.xml.transform.stream.StreamSource
import java.io.FileInputStream
import java.io.IOException

object XsltTransform {
  val FOLDER = "/testcode/xsl/"
}

class XsltTransform {
  @throws[TransformerException]
  def xslt1SafeStaticResource(): Unit = {
    val factory = TransformerFactory.newInstance
    val xslt = new StreamSource(getClass.getResourceAsStream(XsltTransform.FOLDER + "xsl_safe.xsl"))
    val transformer = factory.newTransformer(xslt)
    val text = new StreamSource(getClass.getResourceAsStream(XsltTransform.FOLDER + "input.xml"))
    transformer.transform(text, new StreamResult(System.out))
  }

  @throws[TransformerException]
  def xslt2UnsafeResource(input: String): Unit = {
    val factory = TransformerFactory.newInstance
    val xslt = new StreamSource(getClass.getResourceAsStream(input))
    val transformer = factory.newTransformer(xslt)
    val text = new StreamSource(getClass.getResourceAsStream(XsltTransform.FOLDER + "input.xml"))
    transformer.transform(text, new StreamResult(System.out))
  }

  @throws[TransformerException]
  def xslt3UnsafeResource(input: String): Unit = {
    val factory = TransformerFactory.newInstance
    val xslt = new StreamSource(getClass.getResourceAsStream(XsltTransform.FOLDER + input))
    val transformer = factory.newTransformer(xslt)
    val text = new StreamSource(getClass.getResourceAsStream(XsltTransform.FOLDER + "input.xml"))
    transformer.transform(text, new StreamResult(System.out))
  }

  @throws[TransformerException]
  def xslt4UnsafeResource(input: String): Unit = {
    val factory = TransformerFactory.newInstance
    var in: FileInputStream = null
    try {
      in = new FileInputStream(XsltTransform.FOLDER + input)
      val xslt = new StreamSource(in)
      val transformer = factory.newTransformer(xslt)
      val text = new StreamSource(getClass.getResourceAsStream(XsltTransform.FOLDER + "input.xml"))
      transformer.transform(text, new StreamResult(System.out))
    } catch {
      case e: IOException =>
        e.printStackTrace
    } finally IOUtils.closeQuietly(in)
  }

  @throws[TransformerException]
  def xslt5SafeResource(): Unit = {
    val input = "/safe.xsl"
    val factory = TransformerFactory.newInstance
    var in: FileInputStream = null
    try {
      in = new FileInputStream(XsltTransform.FOLDER + input)
      val xslt = new StreamSource(in)
      val transformer = factory.newTransformer(xslt)
      val text = new StreamSource(getClass.getResourceAsStream(XsltTransform.FOLDER + "input.xml"))
      transformer.transform(text, new StreamResult(System.out))
    } catch {
      case e: IOException =>
        e.printStackTrace
    } finally IOUtils.closeQuietly(in)
  }
}
